stringdtype/stringdtype/src/casts.c

```diff
@@ -42,8 +42,7 @@ string_to_string_resolve_descriptors(PyObject *NPY_UNUSED(self),
                                      npy_intp *view_offset)
 {
     if (given_descrs[1] == NULL) {
-        Py_INCREF(given_descrs[0]);
-        loop_descrs[1] = given_descrs[0];
+        loop_descrs[1] = stringdtype_get_unique_array_descr(given_descrs[0]);
```

Comment:
Mainly thinking.  This situation is interesting because we have 2/3 different situations:
* `arr[...] = arr2` assignments, where the result dtype is "fixed"
* `arr.astype(arr2.dtype)` where the result dtype is requested, but we need a new one.
* `arr.astype(StringDType)` where the result dtype is `NULL`.  I think in this case we can assume a new array is being created.

In the end, we just don't know if we are dealing with an existing allocation or a new one!  One additional caveat is that saying that `view_offset=0` seems only reasonable if we are actually dealing with an existing allocation.

Thoughts (maybe random):
* We could make sure `.astype(dtype)` calls the `get_unique_descr` function or similar?  But:
  * `np.empty()`, etc. it would be nice if the array creation could take care of setting "owned" (and creating a new instance).
  * If `astype()` creates the unique dtype (when passed in), it could indicate to the array creation to use the descriptor verbatim in principle, or would it be necessary to flag that the descriptor is "unowned" and the array creation can just set it to "owned"?
* I was thinking about just using "not owned" but keeping the allocator might work (and then swapping it out).  But that breaks `view_offset=0` semantics, I think.  (i.e. figuring this out in hind-sight).

Have to think about it more, sorry its fuzzy.  Maybe we can also discuss async/sync, but maybe you have some ideas based on this.

Reply:
Hmm, I was under the (mistaken?) impression that we only entered the `given_descrs[1] == NULL` branch if someone does something like `astype(StringDType)`.

I'm having some trouble coming up with examples where this happens. Both scripts in the details blocks below go into the other branch in gdb.

> arr[...] = arr2 assignments, where the result dtype is "fixed"

<details>

```python
from stringdtype import StringDType
import numpy as np

arr = np.array(["hello" * 25, "world" * 25] * 25, dtype=StringDType())

dt = arr.dtype

arr2 = np.array(["foo" * 25, "bar" * 25] * 25, dtype=StringDType())
arr[...] = arr2

print(arr)
assert dt is arr.dtype
```

</details>

> arr.astype(arr2.dtype) where the result dtype is requested, but we need a new one.

<details>

```python

from stringdtype import StringDType
import numpy as np
from numpy._core._multiarray_umath import _get_sfloat_dtype

arr = np.array(["hello" * 25, "world" * 25] * 25, dtype=StringDType(na_object=np.nan))
arr2 = np.array(["hello"*10, "world"*10], dtype=StringDType)
dt = arr2.dtype
arr2 = arr.astype(arr2.dtype)

print(arr2)
assert dt is not arr2.dtype

```

</details>

Reply:
Right, the `NULL` case is probably not very interesting.  And yes both of those cases look identical to resolve dtypes.  Part of the problem may be that we need to distinguish the two cases.
But, I think I thought in the wrong direction.  This may be a problem, but `astype()` currently works like this:
```
# cast user provided dtype, which uses `resolve_descriptors`
new_dtype = cast_dtype(arr.dtype, dtype)
result = np.empty_like(arr, dtype=new_dtype)
result[...] = arr
```
So we we (may) call `resolve_descriptors` twice, but it gets the right descriptor.  And this is not a problem (resolve dtypes always gets the final dtype if it gets one).
But, it may still be important to keep in mind, that in general we have to ensure this.

So back to the multi-step cast:
* We are casting `orig_string -> unicode` and then insert `orig_string -> make_unique(orig_string)` before that.
  * This is because we are creating a new buffer (for silly reasons, but OK).
  * I find it a bit awkward that the second cast works with a different instance then the first.  This works out because `make_unique()` shouldn't do anything weird in this case (even if you broaden its meaning, `orig_string` was already attached to an array after all).  So this is a requirement here that `make_unique` must return something that is OK to read from with the original dtype.  That dtype must have had `make_unique()` called on it, so this is a requirement that `make_unique(make_unique(dtype))` and `make_unique(dtype)` are ABI compatible for *reading*.
    * While we might know we will need a buffer, unfortunately we don't know whether we need a multi-step cast anyway, so I think we cannot move the `make_unique()` to before setting up the first cast :(.
* I am worried a bit because having to insert it here for a buffer creation means that we probably have to do the same thing in other places also.  There shouldn't be a huge number of such buffers being set up, so it isn't a catastrophy.
  * With adapting `asdtype()` it might be that any `resolve_descriptors()` call should never have the identical input/output descriptor, might be a way to add an assert at least?
  * Unfortunately, I am not sure we could use this to decide to return a new one in `resolve_descriptors()` for convenience (to some paths at leasts).  Right now, it enforces identity for same DType casts for the sake of it only.  So we could allow this (and then use the returned one for the buffer), however it would mean that we cannot allow within DType casts to have NumPy set up a multi-step cast behavior for them in the future.  (Which could be neat for complicated unit casts.)

OK, too long... in the end, I am suspecting we really just need it and the `make_unique` constraint is fine in practice.  We *could* yet replace the dtype for the `get_loop()` call, which moves the constraint that `resolve_descriptors()` cannot assume that `make_unique()` won't mutate descriptors before the `get_loop()` call, which is maybe clearer.
The problem with buffer set up might just be minor, but we will have to take a lot more care about cases where we assume dtypes will always be the same.  Two such examples:
* The (fast)take implementation assumes the dtype is identical.  If we assume that make-unique doesn't bother *reading*, all that is needed is inserting the call (or fetching the actual descriptor from the array, but there is the subtlty that our `S0` is barfed as it gets replaced with `S1` by default, so that might need fixing at the same time.)
* The sorting buffer setup needs the call.

Reply:
Hmmm, the `unicode -> string` cast may insert a step also afterwards.  That seems wrong in the other PR?  Because I think you made it `string -> make_unique(string)`, but we must not mutate the result dtype there since it may already be attached to a NumPy array?

Maybe it is correct to move the make unique ahead of time after all?
* If we know we *must* use a buffer, just make the dtype unique for that step.  But for deciding whether to *chain* we must keep using the original dtypes of course!
  * If we must chain anyway due to casting, it shouldn't matter whether the input was "made unique", we replace it again with something that is unique anyway.
  * If the cast can work with the dtype directly, we will still notice the necessary copy step, since we killed the dtype identity ahead of time.

In the end, every time we know we are creating a new buffer, it seems like we want to create a new unique dtype.

I am also wondering whether we should actually indicate if a dtype is made unique for buffering purposes or at array creation.

Reply:
Good call on `sort` and `fasttake` - the existing tests I had for those didn't create any non-short strings so I didn't catch those cases.